traffic_ops/testing/api/v14/user_test.go

```diff
@@ -42,12 +45,59 @@ func CreateTestUsers(t *testing.T) {
 
 		resp, _, err := TOSession.CreateUser(&user)
 		if err != nil {
-			t.Fatalf("could not CREATE user: %v\n", err)
+			t.Errorf("could not CREATE user: %v\n", err)
 		}
 		log.Debugln("Response: ", resp.Alerts)
 	}
 }
 
+func RolenameCapitalizationTest(t *testing.T) {
+
+	var err error
+	var jsonBytes []byte
+
+	user := tc.User{}
+	blob := []byte(`
+	{
+		"username": "test_user",
+		"email": "cooldude6@comcast.net",
+		"fullName": "full name is required",
+		"localPasswd": "better_twelve",
+		"confirmLocalPasswd": "better_twelve",
+		"role": 1,
+		"tenantId": 1
+	}`)
+	err = json.Unmarshal(blob, &user)
+	if err != nil {
+		t.Errorf("could not unmarshal json into struct: %v\n", err)
+	}
+
+	resp, _, err := TOSession.CreateUser(&user)
+	if err != nil {
+		t.Errorf("could not CREATE user: %v\n", err)
+	}
+	jsonBytes, err = json.Marshal(resp)
+	if err != nil {
+		t.Errorf("coult not marshal struct to bytes: %v\n", err)
+	}
+	if bytes.Contains(jsonBytes, []byte("rolename")) {
```

Comment:
I'd maybe invert the logic here and fail the test if it does NOT contain `"roleName": "<insert requested role name here>"`. We could probably "fix" this API by just returning the same data in `"rolename"` _and_ `"roleName"`. That might actually be easier than trying to remember which API returns which form of capitalization. I think that would also mean being able to use the same golang struct for all the users endpoints, kind of like how you changed it originally to embed a struct with the common users fields, which we could probably use here rather than checking the jsonBytes.

I might bring that users API change up on the mailing list to be tackled in a different PR. What do you think?

Reply:
> I think that would also mean being able to use the same golang struct for all the users endpoints, kind of like how you changed it originally to embed a struct with the common users fields, which we could probably use here rather than checking the jsonBytes.

I'm not 100% sure what you're saying here. Currently the same struct is used for all `users/{id}?` endpoints. The `users/current` endpoint isn't affected by the fix and was separate beforehand for other reasons.

> We could probably "fix" this API by just returning the same data in "rolename" and "roleName"
> ...
> I might bring that users API change up on the mailing list to be tackled in a different PR. What do you think?

I think that just makes things potentially more confusing and is unnecessary, but you can ask the mailing list if you want to see other opinions.

Reply:
So we have this `commonUserFields` struct that is embedded in both the structs used for the /users endpoints and the /user/current endpoints. What I'm saying is we could have `roleName` and `roleNameGet` in the `commonUserFields` struct. Then for all the users endpoints we would simply populate both fields with the actual role name. So all the users endpoints would return:
```
...
"rolename": "admin",
"roleName": "admin",
...
```
Then we wouldn't have to special-case the GET endpoint vs the POST/PUT endpoints. Does that make sense?

Reply:
> I don't like the idea of having two role names in part because the user might try to set only one. How would we know which role name to use if we needed to reference it? Were you originally thinking of making a getter and setter surrounding both fields?

That's the thing though -- these are read-only fields. The user _can_ set them in the request, but anything the user sets them to should be ignored. The only _write_ field the API cares about is `role` which is the id of the requested role. The `roleName` fields are only populated after reading the role name from the database to return back to the user.

Come to think of it, the PR as it is now would break any golang clients that hit the `users` endpoint, because the struct field has been renamed `RoleName` -> `RoleNameGET`. Existing clients will expect the `RoleName` field to contain the actual role name, but will instead end up with a nil value which would likely cause a panic. They will have no idea about the new `RoleNameGET` field. So I think the only solution we can provide here is to keep a single `RoleName` field but do custom JSON marshalling so that both `"roleName"` and `"rolename"` get populated with the same value in the JSON. Then the JSON unmarshalling would simply take the one `"roleName"` field as the value for `RoleName`.

This would prevent existing golang clients from breaking, and non-golang clients would be able to read either `"roleName"` or `"rolename"` from the JSON since it will be the same value (since they've likely become accustomed to reading both, depending on the method used).

Also, once we actually fix the golang API to return the proper fields on POST/PUT, we won't end up with a second variation of the special-casing to do `roleName if GET but rolename if PUT/POST` in the other two endpoints as well. The JSON marshalling would just populate both fields under the hood.

Does this make sense? I think it will work and will prevent breaking not only non-golang clients but also golang clients as well.

Reply:
>populate both fields with the actual role name. So all the users endpoints would return:
>
>...
>"rolename": "admin",
>"roleName": "admin",

I'm not a fan of sending unnecessary fields in JSON. It's unnecessary and potentially confusing to users. IMO making the backend a little more complex and having a clean interface, is better than a clean backend and dirty interface.

Which is to say, I think we should use different structs for the different endpoints.

Reply:
You're both right about the problems. The `if bytes.Contains(jsonBytes, []byte("rolename")) {` is error-prone, and renaming a struct field to `RoleNameGET` will break clients. Renaming the struct itself will also break, but it will break their compile and it's easy to fix; whereas the field name breaks at runtime and is very subtle and difficult to track down.

Likewise, sending duplicate fields in the JSON is not only a mess, but will lead to problems with no sane solution. Which value is used if someone posts an object with two different values in `rolename` and `roleName`?

The only solution that solves all these problems, is to separate the structs: `UserGet`, `UserPost`, `UserPutReq`, `UserPutResp`, etc. Because some endpoints are the same and some aren't, type aliases wouldn't be a bad idea either, e.g. `type UserPutResp = UserGet`.

That solves all the problems except (1) the code is a little more complex, and (2) clients may fail to compile and have to change e.g. `tc.User` to `tc.UserGet`. IMO (1) is much less bad than all the other problems, and (2) is inevitable, and a compile-time break is much better than the alternatives.

Reply:
Can't I just keep RoleName as it was and designate it for GET, then make RoleNamePOST instead?

Reply:
Also, I think if we separated things into structs there would only need to be two. One for GET, and one for the others.

Reply:
>What are the existing clients? Wasn't it made within this same release?

Ah, I thought the client functions had been around for a while. If it's the same release, then nothing breaks, great.

>Also, I think if we separated things into structs there would only need to be two. One for GET, and one for the others.

Sure, if they're all the same. The docs weren't clear and I didn't test. I still think type aliases would make things clearer, but I won't insist on it.

Incidentally, we should fix the docs. The docs say /users/id PUT takes "username", but then the example JSON for the same endpoint uses `"userName"`. Same for /users POST.

Reply:
Ok, so the current fix for rolenames works then. I will double check the docs for special casing.